torch/csrc/jit/pybind_utils.h

```diff
@@ -88,82 +97,146 @@ inline c10::optional<TypePtr> unifyOrInitializeType(
+  // Try tensor types
   if (THPVariable_Check(input.ptr())) {
-    auto ten = py::cast<at::Tensor>(input);
-    if (ten.is_sparse()) {
-      AT_ERROR("sparse tensors not supported");
+    auto tensor = py::cast<at::Tensor>(input);
+    if (tensor.is_sparse()) {
+      return MatchTypeReturn("Sparse tensors not supported");
+    }
+    if (tensor.is_mkldnn()) {
+      // mkldnn tensor as opaque tensor doesn't have strides, so we can
+      // not create a CompleteTensorType
+      return MatchTypeReturn(DimensionedTensorType::create(tensor));
+    }
+    return MatchTypeReturn(CompleteTensorType::create(tensor));
+  }
+
+  if (input.is(py::none())) {
+    return MatchTypeReturn("Cannot infer type of a None value");
+  }
+
+  // Try basic types first
+  if (py::isinstance<py::bool_>(input)) {
+    return MatchTypeReturn(BoolType::get());
+  } else if (py::isinstance<py::int_>(input)) {
+    return MatchTypeReturn(IntType::get());
+  } else if (py::isinstance<py::float_>(input)) {
+    return MatchTypeReturn(FloatType::get());
+  } else if (py::isinstance<py::str>(input)) {
+    return MatchTypeReturn(StringType::get());
+  } else if (THPLayout_Check(input.ptr())) {
+    return MatchTypeReturn(IntType::get());
+  } else if (THPDevice_Check(input.ptr())) {
+    return MatchTypeReturn(DeviceObjType::get());
+  } else if (THPDtype_Check(input.ptr())) {
+    return MatchTypeReturn(IntType::get());
+  }
+
+  // Try container types
+  return tryToInferContainerType(input);
+}
+
+inline MatchTypeReturn tryToInferContainerType(py::handle input) {
+  if (THPVariable_Check(input.ptr())) {
+    auto tensor = py::cast<at::Tensor>(input);
+    if (tensor.is_sparse()) {
+      return MatchTypeReturn("Sparse tensors not supported");
     }
-    if (ten.is_mkldnn()) {
+    if (tensor.is_mkldnn()) {
       // mkldnn tensor as opaque tensor doesn't have strides, so we can
       // not create a CompleteTensorType
-      return TypedIValue(ten, DimensionedTensorType::create(ten));
+      return MatchTypeReturn(DimensionedTensorType::create(tensor));
     }
-    return TypedIValue(ten, CompleteTensorType::create(ten));
+    return MatchTypeReturn(CompleteTensorType::create(tensor));
   } else if (six::isTuple(input)) {
-    py::tuple input_tuple = py::cast<py::tuple>(input);
-    Stack s;
-    std::vector<TypePtr> t;
-    s.reserve(input_tuple.size());
-    t.reserve(input_tuple.size());
-    for (py::handle elem : input_tuple) {
-      auto info = toTypedIValue(elem);
-      s.push_back(info.first);
-      t.push_back(info.second);
+    py::tuple tuple = py::cast<py::tuple>(input);
+    std::vector<TypePtr> element_types;
+    element_types.reserve(tuple.size());
+
+    for (py::handle elem : tuple) {
+      auto type_match = tryToInferType(elem);
+      if (type_match.type) {
+        element_types.push_back(*type_match.type);
+      } else {
+        // Forward error message along
+        return type_match.errMsg;
```

Comment:
Can we just return the error here ? ^^ relating to my other comment about nesting the error message

Reply:
It's already returning the error, did you mean something else? What comment about nesting error message?